general/mem_manager.hpp

```diff
@@ -615,41 +751,43 @@ inline Memory<T>::operator const U*() const
 template <typename T>
 inline T *Memory<T>::ReadWrite(MemoryClass mc, int size)
 {
+   const size_t bytes = capacity * sizeof(T);
    if (!(flags & REGISTERED))
    {
       if (mc == MemoryClass::HOST) { return h_ptr; }
-      MemoryManager::Register_(h_ptr, NULL, capacity*sizeof(T),
-                               MemoryType::HOST, flags & OWNS_HOST,
-                               flags & ALIAS, flags);
+      MemoryManager::Register_(h_ptr, nullptr, bytes, h_mt,
+                               flags & OWNS_HOST, flags & ALIAS, flags);
    }
-   return (T*)MemoryManager::ReadWrite_(h_ptr, mc, size*sizeof(T), flags);
+   MemoryManager::CheckHostMemoryType_(h_mt, h_ptr);
```

Comment:
In `Memory::{ReadWrite,Read,Write}`, the calls to `MemoryManager::CheckHostMemoryType_` can be moved inside either `MemoryManager::Register_` or `MemoryManager::{ReadWrite_,Read_,Write_}`.

Reply:
Sure again

Reply:
`done`